internal/dnfjson/dnfjson.go

```diff
@@ -0,0 +1,200 @@
+package dnfjson
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+)
+
+// Solver is configured with a set of repositories and system information in
+// order to resolve dependencies for RPM packages using DNF.
+type Solver struct {
```

Comment:
What we experienced in the past is that if we have one type that is both used as the internal representation and the serialisation format, it makes refactoring very difficult (one has an API guarantee, the other one does not). It also means that as we change the serialisation format in backwards compatible ways, instead of having one place where it is translated to the canonical internal representation (where we could set default values in case we received an old version of the serialised struct), we would have to handle it everywhere.

The real madness though happened when we serialised the same struct in more than one API (weldr + worker + store, for instance), each with slightly different requirements, and on top of that also used the same thing for internal representation. We created ourselves a massive Gordian knot where refactoring anything meant changing everything (and breaking half of it).

I agree that reusing the serialisation format for both sides of a protocol makes sense, but that could be achieved by exposing both the Client and Server side of the protocol from the same package, keeping the serialisation private still, but using the same package on both sides. I imagine something similar could be done for mocking?